### DISCUSSION TITLE
Standard OLED display doesn't work

### DISCUSSION THREAD

**Discussion by ivanvach** [arendst/Tasmota#21121] [Q&A] [answered]
CPU - ESP8266 on Shelly 1
Tasmota binary version    ...display.bin
OLED -  SH1106 1.3" I2C OLED Display Module 128x64 Pixel SH1106

Tasmota commands

I2CScan --> Device found at 0x3C
I2CDriver -> returns drivers 3,4,5,6,38,39,47

Tasmota commands

DisplayMode 0  -> Confirmed
DisplayModel 7 -> Confirmed 

Tasmota command

DisplayText test -> confirmed however the display is dark

What do I do wrong?

Pls. help.

Thanks

**Comment by sfromis** [comment]
Did you power the display on, using the Power command or web UI front page toggle?

Also confirm that the `DisplayAddress` matches, should be 60 (to match the `0x3C`).

**Comment by ivanvach** [comment]
Yes, the address is 60 and Yes the power ON from the web interface did the trick. 

Thank You so very much for the prompt response and invaluable help. You saved the day !

**Comment by amedeodivito** [comment]
from version 14 my SSD1306 OLED display no longer works (version 13.4 works correctly)
I have a d1 mini and the display is connected to pins D1-SCL and D2-SDA
I set OPTIONA3 on the GPIO10 pin but the display remains black
can someone help me?

**Reply by sfromis** [reply]
You also need to have a `display.ini` file, or the content of it by other means.
https://tasmota.github.io/docs/Universal-Display-Driver/#step-5-configure-the-displayini-descriptor-file
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/SH1106_display.ini

With version 14, Tasmota is using the Universal Display driver instead of the older device-specific drivers, hence the need to migrate the setup.

**Reply by amedeodivito** [reply]
"A display.ini file present in the flash file system. preferred option"
does this mean I have to create a file system and put the display.ini file in the DATA folder?
edit:
sto facendo confusione, non so come inserire il file display.ini
p.s.
sorry for my english (I use google translator)

**Reply by sfromis** [reply]
It means that **if** you have a file system (or easily can have it), you can put the `display.ini` at the root of the file system. Otherwise, option 3 may be easier for you. Of course, you should still make sure to be using a binary having the Universal Display driver included, like the precompiled `tasmota-display.bin.gz`.

**Reply by amedeodivito** [reply]
Thank you for your answers and patience
to make the file system management button appear, I loaded tasmota-4M.bin.gz and put the display.ini file of the SSD1306 into the file system but it didn't work, I went back to tasmota-display.bin.gz with optionA-3 and it doesn't work, I uploaded version 13.4 again and everything works, I don't know what to do.
I think I'll have to stick with this version but I'm very sorry

**Reply by sfromis** [reply]
Well, the tasmota-4M build does not support displays, and the tasmota-display build does not have a file system. That's why I suggested that you **not** aim for a file system, but instead use option 3 (from the link I posted), putting the lines from the `display.ini` file in the `Rule3` instead.

Or you can do a custom build yourself, if you want a file system for other purposes, but this should not be needed to support a display.

However, testing with the similar SSD1306 display on ESP8266, I seem to have trouble making it work with the display descriptor in Rule3 instead of the file system. AFAIK, this should work, which implies that there may be a bug here.

**Reply by amedeodivito** [reply]
I'm almost going crazy, I did what you suggested:
OPTION A3
rule3 :H,SSD1306,128,64,1,I2C,3c,*,*,* :S,0,2,1,0,30,20 :I AE D5,80 A8,3F D3,00 40 8D, 14 20.00 A1 C8 DA,12 81.9F D9,F1 DB,40 A4 A6 AF :o,AE :O,AF :A,00,10,40,00,00 :i,A6,A7 #
without inhabiting and enabling it, the result is absolute black.
I also hope it's a bug
However tomorrow I will also try with an esp32.
Thank you so much.

**Reply by sfromis** [reply]
I saw no trouble doing the same test on an ESP32. On ESP8266 it worked for me when using a build with both display support and a file system, putting the descriptor in `display.ini`.

**Reply by arendst** [reply, marked as answer]
Just tried my SH1106 and a rule like 
```
rule3 :H,SH1106,128,64,1,I2C,3c,*,*,* :S,0,2,1,0,30,20 :I AE D5,80 A8,3f D3,00 40 8D,14 20,00 A1 C8 DA,12 81,CF D9F1 DB,40 A4 A6 AF :o,AE :O,AF :A,00,10,40,00,02 :i,A6,A7 #
```
Works OK.

Looking at you rule3 i see some discrepancies which might be causing your issue. Your rule looks like this:
```
rule3 :H,SSD1306,128,64,1,I2C,3c,,,* :S,0,2,1,0,30,20 :I AE D5,80 A8,3F D3,00 40 8D, 14 20.00 A1 C8 DA,12 81.9F D9,F1 DB,40 A4 A6 AF :o,AE :O,AF :A,00,10,40,00,00 :i,A6,A7 #
```
It contains dots where it should have commas. Try the below rule and see if this fixes your issue:
```
rule3 :H,SSD1306,128,64,1,I2C,3c,*,*,* :S,0,2,1,0,30,20 :I AE D5,80 A8,3F D3,00 40 8D,14 20,00 A1 C8 DA,12 81,9F D9,F1 DB,40 A4 A6 AF :o,AE :O,AF :A,00,10,40,00,00 :i,A6,A7 #
```
Pls report your startup log (enable more logging with option 3):
```
00:00:00.001 HDW: ESP8266EX
00:00:00.007 UFS: FlashFS mounted with 1992 kB free
00:00:00.067 CFG: Loaded from File, Count 2255
00:00:00.072 QPC: Count 1
00:00:00.082 DSP: Rule 3 descriptor used
00:00:00.084 I2C: DSP-I2C found at 0x3c
00:00:00.209 DSP: Configured display 'SH1106'
00:00:00.217 SRC: Restart
00:00:00.220 Project tasmota - Wemos6 Version 14.1.0.3(tdisplay)-2_7_7(2024-07-12T10:59:39)
00:00:00.429 WIF: Attempting connection...
```
Notice the DSP messages.

EDIT1: Found my SSD1306 back and currently seems not to work as you noticed too:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.006 UFS: FlashFS mounted with 2000 kB free
00:00:00.062 CFG: Loaded from File, Count 131
00:00:00.067 QPC: Count 1
00:00:00.077 SRC: Restart
00:00:00.079 Project tasmota - Display oled 2 Version 14.1.0.3(tdisplay)-2_7_7(2024-07-12T10:59:39)
00:00:00.181 WIF: Attempting connection...
```
Investigating....

EDIT2: My error. Forgot to add Option A3. Now I see:
```
00:00:00.002 HDW: ESP8266EX
00:00:00.009 UFS: FlashFS mounted with 2000 kB free
00:00:00.066 CFG: Loaded from File, Count 136
00:00:00.072 QPC: Count 1
00:00:00.081 DSP: Rule 3 descriptor used
00:00:00.082 I2C: DSP-I2C found at 0x3c
00:00:00.208 DSP: Configured display 'SSD1306'
00:00:00.216 SRC: Restart
00:00:00.218 Project tasmota - Display oled 2 Version 14.1.0.3(tdisplay)-2_7_7(2024-07-12T10:59:39)
00:00:00.316 CFG: Saved to flash at 3FB, Count 137, Bytes 4096
00:00:00.469 WIF: Attempting connection...
```
Initially no display but after command ``DisplayMode 5`` at least it shows some information. Make sure power is on in main GUI. Also set ``DisplayDimmer 100``. This is my command ``display`` response:
```
{"Display":{"Model":17,"Type":0,"Width":128,"Height":64,"Mode":5,"Dimmer":100,"Size":1,"Font":0,"Rotate":0,"Invert":0,"Refresh":1,"Cols":[21,8],"Rows":8}}
```

EDIT3: Just tried your rule3 and indeed it doesn't work. So the initial issue here is wrong rule3. See above for the correct rule3.

**Reply by amedeodivito** [reply]
ok, now it works
I fixed the error and rule3 should be disabled.
However, I find it rather cumbersome for an inexperienced person like me, I hope that in the future it will be possible to integrate these changes into the precompiled and free rule3 for other purposes
thank you for your precious support
Good day

**Reply by amedeodivito** [reply]
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F8, Count 61
00:00:00.055 QPC: Count 1
00:00:00.057 CFG: CR 379/699, Busy 0
00:00:00.061 TYA: Active=0
00:00:00.064 DSP: Rule 3 descriptor used
00:00:00.065 I2C: DSP-I2C found at 0x3c
00:00:00.191 DSP: Configured display 'SSD1306'
00:00:00.192 SRC: Restart
00:00:00.194 Project tasmota - Tasmota Version 14.1.0(release-display)-2_7_6(2024-06-03T11:53:04)
00:00:00.448 WIF: Attempting connection...

**Reply by sfromis** [reply]
Except for the dots, this equivalent setup works for me on ESP32, but fails on ESP8266:
```
Rule3 :H,SSD1306,128,64,1,I2C,3c,,,* :S,0,2,1,0,30,20 :I AE D5,80 A8,3F D3,00 40 8D,14 20,00 A1 C8 DA,12 81,9F D9,F1 DB,40 A4 A6 AF :o,AE :O,AF :A,00,10,40,00,00 :i,A6,A7 #
```
But I can confirm that using `I2C,3c,*,*,*`, instead of leaving the SCL/SDA pins as empty strings works.

**Comment by Mibeus** [comment]
Hello,

i try compile firmware on github for OLED display, but not possible compile with i2cdriver 3 and 6? How can I do it?

Also when I flash tasmota-display it is without drivers for OLED :(

![compile_problem](https://github.com/user-attachments/assets/b905d207-dd3f-483d-aab1-d83b489280c6)


Thank you and best regards, Michal.

**Reply by sfromis** [reply]
With current Tasmota, you should not try to use the superseded approach of specific drivers for SH1106 etc, but instead use the more general Universal Display Driver, see the migration steps in the docs.
https://tasmota.github.io/docs/Universal-Display-Driver/

**Reply by Mibeus** [reply]
In whitch part this is more simple as set correct driver? 

From this steps I dont understen what I must do on tasmota WEBUI or Platformio or where...
Exist manual with pictures or video?

I need OLED display and DSalias and DS18x20_USE_ID_AS_NAME in one image. But according this steps I am lost :(

With old system was extremly simple. my opinion

**Reply by Mibeus** [reply]
Step 2> I install tasmota-display on ESP8266,
Step 3> skip because I have ESP8266
Step 4 >Set Gpio ports for i2c and one for Option A3 and for DS18b20
Step 5 >Copy Rule3 from arendts
             and now what? What compile????
             `Compile the descriptor into the binary in a section in user_config_override.h under driver 17 (const char)`
           
             
             
![steps](https://github.com/user-attachments/assets/02b7df5c-7da0-4e1b-a421-00b2dece3863)

**Reply by sfromis** [reply]
It is not a question of "more simple" for you, but about the Tasmota software getting simpler and more flexible when not having a bunch of separate drivers for different display boards, like being able to support new display types without extra drivers needing to be created and included.

A standard build with display already can support multiple display types, as long as you carefully follow the steps of getting a display descriptor set up, and an unused pin as `Option A 3`. In my experience, a reasonably painless process, with several ways of including a display descriptor.

If you need a custom build for other reasons, you should make sure to have `USE_UNIVERSAL_DISPLAY` defined in your `user_config_override.h`, instead of trying to include one or more of the no longer available outdated separate display drivers.

**Reply by sfromis** [reply]
You need no compile stuff if you follow the approach with `Rule3`. You could also do a build with a file system.

Also, only one `Option A 3` pin is relevant to have.

BTW. I'd recommend against having the DS18B20 on gpio 0, as any risk of pulling down this pin at boot would mean that Tasmota cannot boot. As a starting point, I like to avoid using gpio 0-3 on those development boards, as they already have a function.

**Reply by Mibeus** [reply]
> You need no compile stuff if you follow the approach with `Rule3`. You could also do a build with a file system.
> 
> Also, only one `Option A 3` pin is relevant to have.

but where should I compile what? In platformIO? Step 2 was to install tasmota-display.

**Reply by Mibeus** [reply]
> You need no compile stuff if you follow the approach with `Rule3`. You could also do a build with a file system.
> 
> Also, only one `Option A 3` pin is relevant to have.
> 
> BTW. I'd recommend against having the DS18B20 on gpio 0, as any risk of pulling down this pin at boot would mean that Tasmota cannot boot. As a starting point, I like to avoid using gpio 0-3 on those development boards, as they already have a function.

Ok, so I set Rule3 and nothing happend..
what means this??Compile the descriptor into the binary in a section in user_config_override.h under driver 17 (const char)

**Reply by sfromis** [reply]
The last is just *another* way of having the display descriptor, not something you need to do. I've not tried that variant and am not sure of the details. Still, if and only if you need a custom build, my preference wold be to do a `tasmota-4M` build to be able to have the display descriptor in the file system. Easier to copy, and does not block one of the rule sets.

With same binary as you, I'm having success with a slightly different display, SSD1306 with a matching `Rule3`. Having `weblog 3` enabled first, at boot I get these confirmation messages:
```
00:00:00.062 DSP: Rule3 descriptor used
00:00:00.064 I2C: SSD1306 found at 0x3c
00:00:00.194 DSP: SSD1306 initialized
```
I saw no such in what you posted earlier, and the two last should also appear with the default `weblog 2`.

Maybe your display descriptor is not "quite right", any errors there could mean that it just does not work. The command `Rule3` should show exactly what you have.

**Comment by Mibeus** [comment]
I will try again in monday with weblog 3.  But I have seconds question
which can make better solution for me. It is possible download older
tasmota version- full source codes? For example v12.x ?

Dňa pi 11. 10. 2024, 16:23 sfromis ***@***.***> napísal(a):

**Reply by sfromis** [reply]
The "Releases" part in the right-hand column of the Github repository front page currently show 107 releases older than latest.

**Reply by Mibeus** [reply]
Thank you, I was before in other page with releases but there was only bin files.

Now I compile own 13.2 version with lcd a and ds18alias and everything works perfect.

Thank you.
